csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,530 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/move_pad.h>
+
+#include <expr_simplifier.h>
+#include <fusion.h>
+#include <ir/builder.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+#include <ops/utils.h>
+#include <transform_replay.h>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+// Simple PadOp is defined as a pad op with:
+// 1. zero pad value; and
+// 2. non-negative pad widths.
+bool isSimplePadOp(PadOp* pad) {
+  if (!simplifyExpr(pad->value())->isZero() &&
+      !simplifyExpr(pad->value())->isFalse()) {
+    return false;
+  }
+  // TODO: we cannot seem to always been able to prove that pad widths are >= 0.
+  // But we know that PadOp + CatOp implies such condition. Adding this
+  // short-cut to ensure that we recognized such pattern.
+  if (pad->out()->uses().size() == 1 && pad->out()->uses()[0]->isA<CatOp>()) {
+    return true;
+  }
+  for (Val* pad_val : pad->getPadWidths()) {
+    if (!simplifyExpr(
+             SimplifyingIrBuilder::geExpr(pad_val, pad->fusion()->zeroVal()))
+             ->isTrue()) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// check if `use` and `p` are the same pad operatoin.
+bool isSamePadOp(Expr* use, PadOp* p) {
+  if (!use->isA<PadOp>()) {
+    return false;
+  }
+
+  auto* use_pad = use->as<PadOp>();
+  if (!simplifyExpr(SimplifyingIrBuilder::eqExpr(use_pad->value(), p->value()))
+           ->isTrue()) {
+    return false;
+  }
+
+  if (use_pad->getPadWidths().size() != p->getPadWidths().size()) {
+    return false;
+  }
+
+  std::vector<int64_t> padded_axes = p->getPaddedAxes();
+  if (use_pad->getPaddedAxes() != padded_axes) {
+    return false;
+  }
+
+  for (auto idx : padded_axes) {
+    if (!simplifyExpr(
+             SimplifyingIrBuilder::eqExpr(
+                 use_pad->getPadWidths(idx).first, p->getPadWidths(idx).first))
+             ->isTrue() ||
+        !simplifyExpr(SimplifyingIrBuilder::eqExpr(
+                          use_pad->getPadWidths(idx).second,
+                          p->getPadWidths(idx).second))
+             ->isTrue()) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// This operation replaces:
+//   CatOp(inputs)
+// with:
+//   BinaryOp(inputs[n-1], BinaryOp(inputs[n-2], BinaryOp(..., BinaryOp[0])...))
+Val* replayCatOpWithBinaryOp(
+    const std::vector<Val*>& inputs,
+    DataType data_type) {
+  // replay `CatOp` with series of BinaryOp instead, since we might have
+  // pushed `PadOp` out and breaking the codegen if `CatOp` remains.
+  Val* res = nullptr;
+  bool is_boolean = isBooleanType(data_type);
+  for (Val* inp : inputs) {
```

Comment:
Nit: assuming `cat` has at least one input, you can simplify the code to something like

```
auto binary_op = is_boolean ? logical_or : add;
Val* res = inputs[0];
for (auto i : c10::irange(1, inputs.size())) {
  res = binary_op(res);
}
return res;
```